src/cluster_import.c

```
@@ -0,0 +1,1927 @@
+            return;
+        }
+
+        listRewind(server.cluster->slot_migration_links, &li);
+        while ((ln = listNext(&li)) != NULL) {
+            slotMigrationLink *link = ln->value;
+            if (!isSlotMigrationLinkInProgress(link)) {
+                continue;
+            }
+            if (link->client) {
+                sendSyncSlotsMessageOnLink(link, "CANCEL");
+                link->client->flag.close_after_reply = 1;
+            }
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_CANCELLED, NULL);
+        }
+        serverLog(LL_NOTICE, "Canceled all in progress slot imports due to operator's request.");
+        addReply(c, shared.ok);
+        return;
+    } else if (!strcasecmp(c->argv[2]->ptr, "link") && c->argc > 3) {
+        slotMigrationLink *link = clusterLookupImportLink(c->argv[3]->ptr);
+        if (!link || !isSlotMigrationLinkInProgress(link)) {
+            addReplyErrorFormat(c, "Link name not found.");
+            return;
+        }
+        if (link->client) {
+            sendSyncSlotsMessageOnLink(link, "CANCEL");
+            link->client->flag.close_after_reply = 1;
+        }
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_CANCELLED, NULL);
+        serverLog(LL_NOTICE,
+                  "Slot import link %.40s to node %.40s (owner of slots [%s]) "
+                  "cancelled due to operator's request.",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        addReply(c, shared.ok);
+        return;
+    } else {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+}
+
+/* Sent by the source to the target after dumping the snapshot in
+ * AOF format. */
+void clusterCommandSyncSlotsSnapshotEof(client *c) {
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!link || link->type != SLOT_MIGRATION_IMPORT) {
+        serverLog(LL_WARNING,
```

I don't think we really need to log this, we should probably return an error to the client and not just kill it as well. If there is some issue in the state machine, we'll just get a disconnect and not a useful error.

---

The reason I don't reply with an error is because both ends of the slot migration link have the read handler installed. So if we reply with an error, it will attempt to parse that as a command. Similar to how replication allows the replica to send REPLCONF ACK over the replication link, and suppresses any responses.

This is how we get away with either end being able to send commands to the other over the same link, but it does prevent responses that are outside of commands.

I have a command - SYNCSLOTS FAIL - which I was using for failure responses. But since the target is the one driving the migration, I don't send it out when the target side fails, since it isn't strictly necessary. Perhaps it would make more sense to send it back in this case so that the error can be seen on either node?

Alternatively - we could use two connections (one incoming and one outgoing) per migration, which would allow us to support responses to each of the SYNCSLOTS messages.

With the current implementation - I was thinking that the operator would see "disconnected" and would investigate the logs on target and see the error log, which would allow them to at least see that something went wrong with the state machine and file an issue. I guess we could also reply with an error, which will produce a log due to the networking.c code that checks that responses aren't sent to the slot migration source.

---

> I guess we could also reply with an error, which will produce a log due to the networking.c code that checks that responses aren't sent to the slot migration source.

This is what I was expecting and makes it a bit easier to debug since you get the error. I suppose I don't feel super strongly one way or the other. It also makes it easier for any type of interoperability system to get a coherent error back if they are trying to interact with it. We (AWS) have one like that for replication but not for slot migration, but in principal a system could try to emulate either side of the slot migration.

---

I've now added this - so we will send:

CLUSTER SYNCSLOTS ESTABLISH ...

And we will enter a state to read out the response before we promote the connection to a client.

It does add some complexity (any time you need to do async IO it takes some code and debugging) but the end result does feel cleaner (and now I can get rid of the additional SYNCSLOTS FAIL command)